packages/worker-relay/src/sqlite/sqlite-relay.ts

```diff
@@ -132,69 +132,82 @@ export class SqliteRelay extends EventEmitter<RelayHandlerEvents> implements Rel
   #insertEvent(db: Database, ev: NostrEvent) {
     if (this.#seenInserts.has(ev.id)) return false;
 
-    const legacyReplacable = [0, 3, 41];
-    if (legacyReplacable.includes(ev.kind) || (ev.kind >= 10_000 && ev.kind < 20_000)) {
-      const oldEvents = db.selectValues("select id from events where kind = ? and pubkey = ? and created <= ?", [
-        ev.kind,
-        ev.pubkey,
-        ev.created_at,
-      ]) as Array<string>;
+    const legacyReplaceableKinds = [0, 3, 41];
+
+    // Handle legacy and standard replaceable events (kinds 0, 3, 41, 10000-19999)
+    if (legacyReplaceableKinds.includes(ev.kind) || (ev.kind >= 10_000 && ev.kind < 20_000)) {
+      const oldEvents = db.selectValues(
+        `SELECT id FROM events WHERE kind = ? AND pubkey = ?`,
+        [ev.kind, ev.pubkey]
+      ) as Array<string>;
+
       if (oldEvents.includes(ev.id)) {
-        // we already have this event, return
+        // Already have this event
         this.#seenInserts.add(ev.id);
-        if (oldEvents.length > 1) {
-          const toDelete = oldEvents.filter(a => a !== ev.id);
-          this.#deleteById(db, toDelete);
-        }
         return false;
       } else {
-        // delete older versions
+        // Delete older events of the same kind and pubkey
         this.#deleteById(db, oldEvents);
       }
     }
+
+    // Handle parameterized replaceable events (kinds 30000-39999)
     if (ev.kind >= 30_000 && ev.kind < 40_000) {
-      const dTag = ev.tags.find(a => a[0] === "d")![1];
+      const dTag = ev.tags.find(a => a[0] === "d")?.[1] ?? "";
+
       const oldEvents = db.selectValues(
-        "select id from events where id in (select id from events, tags where events.id = tags.event_id and tags.key = ? and tags.value = ?)",
-        ["d", dTag],
+        `SELECT e.id
+         FROM events e
+         JOIN tags t ON e.id = t.event_id
+         WHERE e.kind = ? AND e.pubkey = ? AND t.key = ? AND t.value = ?`,
+        [ev.kind, ev.pubkey, "d", dTag]
```

Comment:
Wow, not sure how i missed this being completely wrong, maybe i only worked with events which had unique `d` tags

Reply:
Happens